[testutil] add flake testing marker

pducolin opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

Implements support to mark and optionally skip flaky tests

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

Flaky tests are tests that fail, sometimes. They pass, eventually, on retry. Most of the time, they are an unwanted barrier for dev branches, preventing merge and requiring a manual triggered retry. 

We currently handle flaky tests allowing failure on the job that runs them. This has the downside of allowing failures on all tests running in the same job of a flaky test, hiding actual issues.

We want to improve the current quarantine experience by allowing flagging single go tests as flaky. This is why we are introducing the `flake.Mark` test utils. 

When a test is known as flaky, we will skip it by default on the CI, preventing unwanted CI cost. To keep track of its flake state, we will keep running it on `main` branch, retrying it on failure. If it passes, it will prevent breaking the job it is executed on. If it does not pass after retries, it is not a flaky, rather a broken tests and will be disabled. This is similar to what we have today, improving the granularity of tests that are allowed to fail.

#### ℹ Note

In a follow up PR we will skip flakes in dev pipelines and retry flake failures on main pipelines.

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

Do you have a flaky test ? Mark it with `flake.Mark`. 

By default the marker sends a log with a special message that marks it as known flake.

Optionally with **skip flake**, the CI skips the test as soon as the marker is called. Use `--skip-flake` or set `GO_TEST_SKIP_FLAKE` environment variable to true.

See examples in [flake_test.go](https://github.com/DataDog/datadog-agent/blob/pducolin/APL-2357-allow-tagging-flaky-tests/pkg/util/testutil/flake/flake_test.go)

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

pducolin committed:
[testutil] add flake testing wrapper

Implements support to allow or skip flaky tests

pgimalac commented on pkg/util/testutil/flake/flake.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+// Wrap wraps a testing.TB to allow flaky tests to fail or skip them.
+func Wrap(t testing.TB) testing.TB {
```
I like the idea of being able to easily mark a test as flaky, but I think it should be a one liner, eg. just add a `t = flaky.Wrap(t)` at the start of the test.

The current implementation doesn't allow you to do that, so it doesn't prevent you from using the initial `testing.T`, since the wrapper doesn't provide all methods of `testing.T` (it's a `testing.TB`).

I checked in the repo and
- `testing.T` is used 7868 times
- `testing.B` is used 591 times
- `testing.TB` is used 27 times
- `testing.F` is used 25 times

Considering that, I think having a specific implementation of `Wrap` for `testing.T` which returns a `testing.T` and allows overriding the variable in a single line would be valuable.

Sadly you can't have an embedded field be a type parameter so you can't do the following, but I still think having a duplicate implementation specifically for `T` could be worth it. 
```go
type allowflakeTesting[T testing.TB] struct {
	T
}
```

What do you think ?

pducolin replied:
I see the limitation, we would like to return `testing.T` from `Wrap`, while catching calls to `Fail` and other `testing.T`. It is an anti-pattern to return an interface in Go.

We could always return the `allowFlakeTesting` struct, storing the mode in the struct's state: this was my initial dev implementation. Having a `mode=[None,Skip,Allow]` based on the flags value. It means changing the override implementation to call `t.<Func>` on `None|Skip` mode and calling the current code on `Allow` mode.

I thought that was too complex, and it does not solve the different type issue that prevents reusing the same variable name :( 

Also note that there is no way to support `t.Run`, without reimplementing it. Unfortunately `t.Testing` is not meant for such usage. It is a struct, not an interface, so we cannot wrap it easily and transparently.

We could improve devx with an internal tool to flag a single test as flaky, that would take care of replacing `t` with `ft`

pducolin replied:
Resolving this conversation as I changed the implementation

pgimalac commented on pkg/util/testutil/flake/flake_test.go:
```diff
@@ -0,0 +1,140 @@
+func wrapAndRunSubFlakyTest(t *mockTesting) {
+	t.Helper()
+	for i := 0; i < 100; i++ {
+		t.Run("sub flake test", func(t *testing.T) {
+			ft := Wrap(t)
+			assert.Equal(ft, "heads", flipCoin())
+		})
```
This is a good example of my previous comment: when you add the call to `Wrap` you need to either rename all uses of `t` to `tb`, or rename the argument `t` to something else (but in that case you'll get errors if using methods of `testing.T` which are not in `testing.TB`).

ogaca-dd commented:
> allow flake failures intercepts failures and skip the failing test, logging the error. Use --allow-flake-failure flag or set ALLOW_FLAKE_FAILURE environment variable to true

What is the usage of this feature? I am asking because if it is not strictly needed an easy solution would be to use something like
```go
func TestFlaky(t *testing.T) {
	FlakyTest(t)
	require.True(t, false)
}

func FlakyTest(t *testing.T) {
	if strings.ToLower(os.Getenv("SKIP_FLAKE")) == "true" {
		t.Skip("Skipping flaky test")
	}
}
```

ogaca-dd commented on pkg/util/testutil/flake/flake_test.go:
```diff
@@ -0,0 +1,140 @@
+	return mt.failCallCount
+}
+
+var trueValue = true
```
```suggestion
const trueValue = true
```

pducolin replied:
I use it to set a pointer, go does not allow getting the address of a const  https://stackoverflow.com/questions/35146286/find-address-of-constant-in-go

ogaca-dd commented on pkg/util/testutil/flake/flake_test.go:
```diff
@@ -0,0 +1,140 @@
+
+func wrapAndRunFlakyTest(t *mockTesting) {
+	t.Helper()
+	wg := sync.WaitGroup{}
+	// testing.T.Skip() calls runtime.Goexit() which terminates the goroutine
+	// run the test in a separate goroutine to avoid terminating `TestFlake` test
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		ft := Wrap(t)
+		for i := 0; i < 100; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				coin := flipCoin()
+				assert.Equal(ft, "heads", coin)
+			}()
+		}
+	}()
+	wg.Wait()
```
You don't need two layers of goroutine and the following code works.
```go
	ft := Wrap(t)
	for i := 0; i < 100; i++ {
		wg := sync.WaitGroup{}
		wg.Add(1)
		go func() {
			defer wg.Done()
			coin := flipCoin()
			assert.Equal(ft, "heads", coin)
		}()
		wg.Wait()
	}
```
By having a single goroutine at a time (Note the location of `wg.Wait()`) you can remove `mutex`.
What do you think?

pducolin replied:
I need to have two layers as I need `ft := Wrap(t)` to happen in a goroutine. `t.Skip` calls `runtime.Goexit()` underneath, that terminates the goroutine where it is called. 

Since I use this same function with `skip` mode, if I call `ft := Wrap(t)` from the test runner thread, it will terminate the actual test, `TestFlake`. 

I know it's confusing, there is the comment at the beginning of the first call to a goroutine, but I should make it more explicit.

Note also this is also for meta testing purpose, actual tests will wrap on the main thread and stop the main test on flake

pducolin replied:
Without the first goroutine level, on the test with `skipTest` true we call skip on the main thread and skip the actual test, not just the metatest

ogaca-dd reviewed (commented):
Great PR!!!

pducolin commented on pkg/util/testutil/flake/flake.go:
```diff
@@ -0,0 +1,91 @@
+	if *allowFlakeFailure {
+		return true
+	}
+	if os.Getenv("ALLOW_FLAKE_FAILURE") == "true" {
```
Thinking we could parse it to bool to handle also `1`

pducolin replied:
Should be handled now

pducolin commented on pkg/util/testutil/flake/flake_test.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+func (mt *mockTesting) Skip(_ ...any) {
+	func() {
```
It's to unlock the mutex once the counter is set.

pgimalac replied:
You could not defer the call to `Unlock` I guess

pducolin replied:
Yes, could call it explicitly in this 3 liner-function

pducolin replied:
Could I address this in a follow up ?

pgimalac replied:
It's a style suggestion so whenever you want, it doesn't matter much IMO

pducolin committed:
[util] allow marking tests as flakes

Allows both skipping them through a flag and / or environment variable, or retry them to have a passing result. They would still report as flaky, but would not impact neighbours

pducolin commented:
On flaky test management, found this issue that is accepted and supposedly under dev on the go repo: https://github.com/golang/go/issues/62244

The discussion is super interesting, and made me change my mind in the way I was thinking handling flakiness. I still want to skip them by default, but when we run them on main we should retry them until the pass, max N time - N to be defined. Then the reporting code should catch the marker log, report them as flaky with success after M attempts. If a test fails more than N times in a row, it is not flaky, it is broken, and should be disabled and fixed as it brings no value.

pducolin committed:
[codeowner] update codeowners for plkg/util/testutil/flake to agent platform

pducolin committed:
[flake] parse env var to boolean

ogaca-dd commented on pkg/util/testutil/flake/flake.go:
```diff
@@ -38,8 +39,9 @@ func shouldSkipFlake() bool {
 	}
-	if os.Getenv("GO_TEST_SKIP_FLAKE") == "true" {
-		return true
+	shouldSkipFlakeVar, err := strconv.ParseBool(os.Getenv("GO_TEST_SKIP_FLAKE"))
```
👍

pducolin committed:
[flake] update package name

pducolin commented on pkg/util/testutil/flake/flake.go:
```diff
@@ -0,0 +1,47 @@
+// Use flags `allow-flake-failure` and `skip-flake` to control the behavior, or set the environment variables `ALLOW_FLAKE_FAILURE` and `SKIP_FLAKE`.
+// If both `allow-flake-failure` and `skip-flake` are set, `skip-flake` takes precedence.
+// Flags take precedence over environment variables.
+// Subtests from [testing.T.Run](https://pkg.go.dev/testing#T.Run) should be wrapped with [flake.Wrap](https://pkg.go.dev/github.com/DataDog/datadog-agent/pkg/util/testutil/flake#Wrap) to inherit the behavior.
```
Updated

pducolin commented:
/merge
